### PR TITLE
Stateless builder

### PR DESCRIPTION
**Pull request by ferranbt**
## 📚 References

---

* [x] I have seen and agree to CONTRIBUTING.md

### CODE REVIEW

File: core/vm/contracts_suave.go [hunk 1]
```diff
@@ -196,3 +197,20 @@ func (s *suaveRuntime) doHTTPRequest(request types.HttpRequest) ([]byte, error)
 	}
 	return data, nil
 }
+
+func (s *suaveRuntime) newBuilder() (string, error) {
```
Comment (dmarzzz) [inline core/vm/contracts_suave.go:201]: would be interesting to think about two kettles sharing the same session ID and doing a `newBuilder(existingSessionID string)` or `connectBuilder`. We don't need to consider for this work just mentioning if it inspires anything again!

File: suave/builder/api/api_server.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+func NewServer(s sessionManager) *Server {
+	api := &Server{
+		sessionMngr: s,
+	}
+	return api
+}
+
+type Server struct {
+	sessionMngr sessionManager
```
Comment (dmarzzz) [inline suave/builder/api/api_server.go:23]: should we have a mutex on this? seems possible you could call start a session from another SUAPP before the last new session call has been completed?

File: suave/builder/api/api_server.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+func (s *Server) NewSession(ctx context.Context) (string, error) {
```
Comment (dmarzzz) [inline suave/builder/api/api_server.go:26]: wonder if we should consider some sort of sane session cap, additionally, should we add metrics to this maybe so we can observe?

File: suave/builder/builder.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+type builderConfig struct {
+	preState *state.StateDB
+	header   *types.Header
+	config   *params.ChainConfig
+	context  core.ChainContext
+}
+
+func newBuilder(config *builderConfig) *builder {
```
Comment (dmarzzz) [inline suave/builder/builder.go:28]: @Ruteri curious your thoughts on if this is the best way to set up builder, I remember something non-obvious about reverts changing in builder recently so double check would be appreciated

File: suave/builder/session_manager.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	parent := s.blockchain.CurrentHeader()
+
+	chainConfig := s.blockchain.Config()
+
+	header := &types.Header{
+		ParentHash: parent.Hash(),
+		Number:     new(big.Int).Add(parent.Number, common.Big1),
+		GasLimit:   core.CalcGasLimit(parent.GasLimit, s.config.GasCeil),
+		Time:       1000,             // TODO: fix this
+		Coinbase:   common.Address{}, // TODO: fix this
```
Comment (dmarzzz) [inline suave/builder/session_manager.go:77]: hmm yeah interesting to think what you would set this to. Two ideas come to mind:
- set to '0x + keccak256("SUAVE")` so its consistent for users 
- allow for the ability to set it. I think there may be some scenarios where you if your a builder who also has a contract controlled by your address that does something different based on profit in a block, then you would want to be able to set this. Happy to spell this out more but I think it's desirable.

File: core/vm/contracts_suave.go (hunk 1 above)
Comment (dmarzzz) [reply core/vm/contracts_suave.go:201]: permisisons around this has some caution signs for me, you could mess with someones simulation if you knew their session ID if done naively

File: suave/gen/suave_spec.yaml [hunk 6]
```diff
@@ -218,3 +236,20 @@ functions:
       fields:
         - name: response
           type: bytes
+  - name: newBuilder
+    address: "0x0000000000000000000000000000000053200001"
+    output:
+      fields:
+        - name: id
+          type: string
```
[the remark this answers is not in the document]
Comment (ferranbt) [reply suave/gen/suave_spec.yaml:244]: The session id is a `uuid` identifier. It gets rendered as a string.

File: suave/builder/session_manager.go (hunk 5 above)
Comment (ferranbt) [reply suave/builder/session_manager.go:77]: Agree, I think some of these params should be configurable when you start a session like the `extra` field.

File: suave/builder/api/api_server.go (hunk 2 above)
Comment (ferranbt) [reply suave/builder/api/api_server.go:23]: The locks are on `sessionManager`.

File: suave/sol/libraries/Suave.sol [hunk 7]
```diff
@@ -34,6 +34,19 @@ library Suave {
         bytes body;
     }
 
+    struct SimulateTransactionResult {
+        uint64 egp;
```
[the remark this answers is not in the document]
Comment (ferranbt) [reply suave/sol/libraries/Suave.sol:38]: I would see that included as some type of getBalance call to the builder.

File: suave/builder/api/api_server.go (hunk 3 above)
Comment (ferranbt) [reply suave/builder/api/api_server.go:26]: We could address this in a followup PR.

**Review by dmarzzz — APPROVED** [review]
LSTM

File: suave/builder/api/api_server.go (hunk 3 above)
Comment (lthibault) [reply suave/builder/api/api_server.go:26]: #165

**Pull request merged**